src/gfnff/gfnff_ini2.f90

```diff
@@ -774,85 +775,110 @@ subroutine gfnff_hbset(n,at,xyz,topo,neigh,nlist,hbthr1,hbthr2)
+               ! get adjustet iTr -> for use of neigh% distances and bpair with two shifted atoms
+               iTrDum=neigh%fTrSum(neigh%iTrNeg(iTri),iTrj)
+               ! iTrDum=-1 is valid here because we are only interested if ij is a bond (ijnonbond)
+               ! However, iTrDum is not and should not be used as an index without excluding -1 value
+               if(iTrDum.gt.neigh%nTrans.or.iTrDum.lt.-1.or.iTrDum.eq.0) cycle
+               vec_ab = (xyz(:,i)+neigh%transVec(:,iTri))-(xyz(:,j)+neigh%transVec(:,iTrj))
+               rab = mctc_dot(vec_ab, vec_ab)  ! square of distance AB
+               if(rab.gt.hbthr1) cycle
+               ! check if ij bonded
+               if(iTrDum.le.neigh_numctr.and.iTrDum.gt.0) then
+                  ijnonbond=neigh%bpair(j,i,iTrDum).ne.1
+               else
+                  ! i and j are not in neighboring cells
+                  ijnonbond=.true.
+               endif
+               ! loop over relevant H atoms
+               do k=1,topo%nathbH
+                  free=.true. ! tripplet not assigned yet
+                  nh  =topo%hbatHl(1,k) ! nh always in central cell
+                  ! distances for non-cov bonded case
+                  vec_ih = xyz(:,nh)-(xyz(:,i)+neigh%transVec(:,iTri))
+                  rih = mctc_dot(vec_ih, vec_ih)  ! square of distance iH
+                  vec_jh = xyz(:,nh)-(xyz(:,j)+neigh%transVec(:,iTrj))
+                  rjh = mctc_dot(vec_jh, vec_jh)  ! square of distance jH
+                  ! check if i is the bonded A
+                  if(iTri.le.neigh_numctr) then ! nh is not shifted so bpair works without adjustment
+                     if(neigh%bpair(i,nh,iTri).eq.1.and.ijnonbond) then
+                        !$omp atomic capture
+                        nlist%nhb2=nlist%nhb2+1
+                        nlist_nhb2 = nlist%nhb2
+                        !$omp end atomic
+                        nlist%hblist2(1,nlist_nhb2)=i
+                        nlist%hblist2(2,nlist_nhb2)=j
+                        nlist%hblist2(3,nlist_nhb2)=nh
+                        nlist%hblist2(4,nlist_nhb2)=iTri
+                        nlist%hblist2(5,nlist_nhb2)=iTrj
+                        free=.false. ! not available for nhb1 !!!
+                     endif
+                  endif
+                  ! check if j is the bonded A
+                  if(iTrj.le.neigh_numctr.and.free) then
+                     if(neigh%bpair(j,nh,iTrj).eq.1.and.ijnonbond) then
+                        !$omp atomic capture
+                        nlist%nhb2=nlist%nhb2+1
+                        nlist_nhb2 = nlist%nhb2
+                        !$omp end atomic
+                        nlist%hblist2(1,nlist_nhb2)=j
+                        nlist%hblist2(2,nlist_nhb2)=i
+                        nlist%hblist2(3,nlist_nhb2)=nh
+                        nlist%hblist2(4,nlist_nhb2)=iTrj
+                        nlist%hblist2(5,nlist_nhb2)=iTri
```

Comment:
Here you may have data race with nlist_nhb2 variable. On x86/x86_64 it works fine, but AArch64 can say "Hi" for you :)